modules/images/webp-uploads/load.php

```diff
@@ -11,42 +11,92 @@
 /**
  * Hook called by `wp_generate_attachment_metadata` to create the `sources` property for every image
  * size, the sources' property would create a new image size with all the mime types specified in
- * `webp_uploads_valid_image_mime_types`. If the original image is one of the mimes from
- * `webp_uploads_valid_image_mime_types` the image is just added to the `sources` property and  not
- * created again. If the uploaded attachment is not a valid image this function does not alter the
- * metadata of the attachment, on the other hand a `sources` property is added.
+ * `webp_uploads_get_supported_image_mime_transforms`. If the original image is one of the mimes from
+ * `webp_uploads_get_supported_image_mime_transforms` the image is just added to the `sources` property and not
+ * created again. If the uploaded attachment is not a supported mime by this function, the hook does not alter the
+ * metadata of the attachment. In addition to every single size the `sources` property is added at the
+ * top level of the image metadata to store the references for all the mime types for the `full` size image of the
+ * attachment.
  *
  * @since 1.0.0
  *
  * @see   wp_generate_attachment_metadata()
- * @see   webp_uploads_valid_image_mime_types()
+ * @see   webp_uploads_get_supported_image_mime_transforms()
  *
  * @param array $metadata      An array with the metadata from this attachment.
  * @param int   $attachment_id The ID of the attachment where the hook was dispatched.
  * @return array An array with the updated structure for the metadata before is stored in the database.
  */
 function webp_uploads_create_sources_property( array $metadata, $attachment_id ) {
-	// Make sure we have some sizes to work with, otherwise avoid any work.
-	if ( empty( $metadata['sizes'] ) || ! is_array( $metadata['sizes'] ) ) {
-		return $metadata;
-	}
 	// This should take place only on the JPEG image.
 	$valid_mime_transforms = webp_uploads_get_supported_image_mime_transforms();
-
 	// Not a supported mime type to create the sources property.
 	$mime_type = get_post_mime_type( $attachment_id );
-	if ( ! isset( $valid_mime_transforms[ $mime_type ] ) ) {
+	if ( ! isset( $valid_mime_transforms[ $mime_type ] ) || ! is_array( $valid_mime_transforms[ $mime_type ] ) ) {
```

Comment:
The `! is_array` check is a bit over-aggressive to add. If it wasn't an array that would be invalid, and in that case the developer using the filter would have clearly done something wrong. If you prefer to validate this, it should happen in a centralized way in `webp_uploads_get_supported_image_mime_transforms()`, so that it doesn't need to be checked after every call to the function.

Reply:
> The ! is_array check is a bit over-aggressive to add. If it wasn't an array that would be invalid, and in that case the developer using the filter would have clearly done something wrong. 

Right, however, if we let them break it, the `foreach` loop won't be executed and triggered an error we can let the error happen in that case or just ignore and not process any invalid value to the filter. I'm more inclined towards the second but let me know if you think otherwise.

> If you prefer to validate this, it should happen in a centralized way in webp_uploads_get_supported_image_mime_transforms(), so that it doesn't need to be checked after every call to the function.

We do the validation in order to allow running in this function as well specifically here:

- https://github.com/WordPress/performance/blob/223137d6190de58762ec49dba720a4737a44bb99/modules/images/webp-uploads/load.php#L72


However I do see an inconsistency over here: 

- https://github.com/WordPress/performance/blob/223137d6190de58762ec49dba720a4737a44bb99/modules/images/webp-uploads/load.php#L138

Where we are assuming the value is an `array` if is defined, so I agree we should standardized that condition, still I'm more inclined towards making sure the value is a valid array but let me know if you think otherwise.

Reply:
> Right, however, if we let them break it, the `foreach` loop won't be executed and triggered an error we can let the error happen in that case or just ignore and not process any invalid value to the filter. I'm more inclined towards the second but let me know if you think otherwise.

I agree we should ignore invalid values and not cause an error. I'd say that should happen by sanitizing the value right after the filter though, within `webp_uploads_get_supported_image_mime_transforms()`. That ensures the return value of the function is reliable, without requiring any calling code to check itself. In other words, `webp_uploads_get_supported_image_mime_transforms()` should strip any array keys where the value is `! is_array`.

Reply:
> I agree we should ignore invalid values and not cause an error. I'd say that should happen by sanitizing the value right after the filter though, within webp_uploads_get_supported_image_mime_transforms(). That ensures the return value of the function is reliable, without requiring any calling code to check itself. In other words, webp_uploads_get_supported_image_mime_transforms() should strip any array keys where the value is ! is_array.

Great point, let me adjust that one.